express/src/controllers/manage-account.ts

```diff
@@ -32,9 +35,119 @@ export const listServices = async function (req: Request, res: Response) {
+    const serviceId = req.params.serviceId;
+    const s4: SelfServiceServicesService = req.app.get("backing-service");
+    const client = (await s4.listClients(serviceId, req.session.authenticationResult?.AccessToken as string))[0];
+    const selfServiceClientId = client.dynamoId;
+    const authClientId = client.authClientId;
+    const serviceName = client.serviceName;
+
+    res.render("service-details/client-details.njk", {
+        clientId: authClientId,
+        selfServiceClientId: selfServiceClientId,
+        serviceId: serviceId,
+        serviceName: serviceName,
+        updatedField: req.session.updatedField,
+        redirectUrls: client.redirectUris.join(" "),
+        userAttributesRequired: client.scopes.join(", "),
+        userPublicKey: client.publicKey == defaultPublicKey ? "" : client.publicKey,
+        postLogoutRedirectUrls: client.logoutUris.join(" "),
+        urls: {
+            // TODO changeClientName is currently not used
+            changeClientName: `/change-client-name/${serviceId}/${selfServiceClientId}/${authClientId}?clientName=${encodeURI(
+                client.clientName
+            )}`,
+            changeRedirectUris: `/change-redirect-uris/${serviceId}/${selfServiceClientId}/${authClientId}?redirectUris=${encodeURI(
+                client.redirectUris.join(" ")
+            )}`,
+            changeUserAttributes: `/change-user-attributes/${serviceId}/${selfServiceClientId}/${authClientId}?userAttributes=${encodeURI(
+                client.scopes.join(" ")
+            )}`,
+            changePublicKey: `/change-public-key/${serviceId}/${selfServiceClientId}/${authClientId}?publicKey=${encodeURI(
+                client.publicKey
+            )}`,
+            changePostLogoutUris: `/change-post-logout-uris/${serviceId}/${selfServiceClientId}/${authClientId}?redirectUris=${encodeURI(
+                client.logoutUris.join(" ")
+            )}`
+        }
+    });
+
+    // TODO we need to use a flash message package for Express
+    req.session.serviceName = serviceName;
+    req.session.updatedField = undefined;
+};
+
+export const showPrivateBetaForm = async function (req: Request, res: Response) {
+    res.render("service-details/private-beta.njk", {
+        serviceId: req.params.serviceId,
+        selfServiceClientId: req.params.selfServiceClientId,
+        clientId: req.params.clientId,
+        serviceName: req.session.serviceName,
+        emailAddress: req.session.emailAddress
+    });
+};
+
+export const processPrivateBetaForm = async function (req: Request, res: Response) {
+    const yourName = req.body.yourName;
+    const department = req.body.department;
+    const serviceName = req.body.serviceName;
+    const emailAddress = req.session.emailAddress;
+    const serviceId = req.params.serviceId;
+    const selfServiceClientId = req.params.selfServiceClientId;
+    const clientId = req.params.clientId;
+    const errorMessages = new Map<string, string>();
```

Comment:
I thought we'd banished these Maps and were onto the macro stuff?

Reply:
Yes a bit further down `errorMessages: Object.fromEntries(errorMessages)`
This is just to simplify the logic in that function for now. The controllers and validation will need more refactoring but that's for another story. This is fine for the MVP